### PR TITLE
fix: @rematch/select typescript plugin compatibility

### PR DESCRIPTION
**Pull request by semoal**
### TODO:
- [x] is it a good practice create a createModelWithSelectors<RootModel>()({})? or could we extends native createModel for plugins?

- selectors on model has three props, we must guarantee it doesn't have errors:
	- [X]  selectors: (slice) => ...
	- [ ]  selectors: (slice, createSelector) => ...
	- [ ]  selectors: (slice, createSelector, hasProps) => ...

- [x] store.select is not autocompleted
```jsx
const selection = store.select(models => ({
  // todo: this isn't getting autocompleted
  total: models.cart.total,
}))
```

- [x] second param shouldn't be required, or yes?.. take a look at reselect examples
```jsx
const mapState = (state: RootState) => ({
	settingsState: state.settings,
  loadingState: state.loading,
  playersState: state.players,
  // todo: props shouldn't be required
  ...selection(state, null)
})
```

- [x] Props for class and hooks, should be filled...
```jsx
// todo: we must type the props autocomplete of the selector new values
type Props = StateProps & DispatchProps
```
- [x] Props for class and hooks, should be filled...
```jsx
// todo: we must type the props autocomplete of the selector new values
type Props = StateProps & DispatchProps
```

### CODE REVIEW

File: packages/select/src/index.ts [hunk 1]
```diff
@@ -64,32 +74,80 @@ const validateConfig = (config: SelectConfig): void => {
...
 			throw new Error(
 				`Selector (${model.name}/${selectorName}) must be a function`
 			)
 		}
 	}
 }
 
-const createSelectPlugin = (config: SelectConfig = {}): Plugin => {
+/**
+ * Todo: could we extend types of standard createModel?
```
Comment (semoal) [inline packages/select/src/index.ts:86]: @tianzhich

File: packages/select/src/types.ts [hunk 2]
```diff
@@ -66,15 +76,34 @@ export type RematchSelect<M extends object = Models, RootState = any> = ((
...
+		state: S
+		reducers?: R
+		selectors?: ModelSelectorsConfig<S>
+		baseReducer?: BR
+		effects?: E
 	}
 
-	interface RematchStore<TModels> {
-		select: RematchSelect<TModels, RematchRootState<TModels>>
+	export interface RematchStore<
```
Comment (tianzhich) [inline packages/select/src/types.ts:101]: why copy the same RematchStore in select?

File: packages/select/src/types.ts [hunk 3]
```diff
@@ -66,15 +76,34 @@ export type RematchSelect<M extends object = Models, RootState = any> = ((
...
 	> {
+		name?: string
+		state: TState
 		selectors?: ModelSelectorsConfig<TState>
+		reducers?: ModelReducers<TState>
+		baseReducer?: ReduxReducer<TBaseState>
+		effects?: ModelEffects<TModels> | ModelEffectsCreator<TModels>
+	}
+
+	export interface ModelResult<S, R, BR, E> {
```
Comment (tianzhich) [inline packages/select/src/types.ts:92]: what is the ModelResult used for?

File: packages/select/src/types.ts [hunk 4]
```diff
@@ -66,15 +76,34 @@ export type RematchSelect<M extends object = Models, RootState = any> = ((
 	StoreSelectors<RootState>
 
 declare module '@rematch/core' {
-	interface Model<
+	export interface Model<
```
Comment (tianzhich) [inline packages/select/src/types.ts:79]: I think we don't need to complete the props of Model, these props can be combined when Model definition is used

File: packages/select/src/index.ts (hunk 1 above)
Comment (tianzhich) [reply packages/select/src/index.ts:86]: I was thinking if we create a modelCreator for each plugin for that purpose. How can we use them together?
Could we add a loose index type for `createModel` just like this:

```ts
export const createModel: <
	RM extends Models<RM>
>() => <
	R extends ModelReducers<S>,
	BR extends ReduxReducer<BS>,
	E extends ModelEffects | ModelEffectsCreator<RM>,
	S,
	BS = S
>(mo: {
	name?: string
	state: S
	reducers?: R
	baseReducer?: BR
	effects?: E
	[index: string]: any
}) => {
	name?: string
	state: S
	reducers: R
	baseReducer: BR
	effects: E
        [index: string]: any
} = () => (mo): any => {
	const { reducers = {}, effects = {} } = mo

	return {
		...mo,
		reducers,
		effects,
	}
}
```

So our users can add their own model configs if they needed when creating a new plugin. But cause the type is any, so they need to use `as` to get type completion.

Comment (tianzhich) [reply packages/select/src/index.ts:86]: Or we can collect their config into a prop called `configs` so we can pass the type param to get the completion.

File: packages/select/src/types.ts (hunk 2 above)
Comment (semoal) [reply packages/select/src/types.ts:101]: i forgot to add the `RematchSelect` type, that's why it's copied

`		select: RematchSelect<TModels, RematchRootState<TModels>>	`

File: packages/select/src/types.ts (hunk 3 above)
Comment (semoal) [reply packages/select/src/types.ts:92]: for nothing, on a first instance i've tried to use it in createModel in core, and extend the type here with selectors

File: packages/select/src/types.ts (hunk 4 above)
Comment (semoal) [reply packages/select/src/types.ts:79]: could you give me an example? i'm not sure how to fill 

`selectors?: ModelSelectorsConfig<TState>`, for the model

File: packages/select/src/index.ts (hunk 1 above)
Comment (semoal) [reply packages/select/src/index.ts:86]: like `createModel<RootModel, ExtraSelectorsModel<RootModel>>()({})`?

File: packages/select/package.json [hunk 5]
```diff
@@ -24,7 +24,8 @@
     "src"
   ],
   "dependencies": {
-    "reselect": "^4.0.0"
+    "reselect": "^4.0.0",
+    "redux": "^4.0.0"
```
[the remark this answers is not in the document]
Comment (semoal) [reply packages/select/package.json:28]: yes, redux should be in peerDeps

File: packages/select/src/types.ts (hunk 2 above)
Comment (semoal) [reply packages/select/src/types.ts:81]: when select plugin it's inserted the store contains a new prop `store.select()`

```jsx
const selection = store.select(models => ({
  // todo: this isn't getting autocompleted
  total: models.cart.total,
}))

const mapState = (state: RootState) => ({
	settingsState: state.settings,
  loadingState: state.loading,
  playersState: state.players,
  // todo: 2n param props shouldn't be required
  ...selection(state, null)
})
```

File: packages/select/src/index.ts (hunk 1 above)
Comment (tianzhich) [reply packages/select/src/index.ts:86]: I am not sure hot to extend native createModel for plugins. If we want the automation from selectors just like effects, seems the only way to work is yours. But if we have another plugin which also need model config. Are we create a helper `createModelWithXXX`? How could we use them together?

Or we can tell our plugin contributers to modify `createModel` in core and add these config definitions, if needed.

**Comment by semoal** [comment]
@tianzhich let's see if we can give this pull request a push/sprint, at least some minimal features like slice it's already done. We should decide createModelSelectors if is interesting or we should extends createModel({}) to something else..

@pjamrozowicz If you have some free time your opinion or help will be really appreciated, this is the only pending task from your migration to ts :)

**Comment by tianzhich** [comment]
@semoal yes, we can give a sprint.

And I used an overload for `createModel` just like what I did for `connect` function of `react-redux`, which was more elegant than `createModelSelectors`. So we can add extra plugin configs efficiently.

File: packages/select/src/types.ts (hunk 2 above)
Comment (semoal) [reply packages/select/src/types.ts:84]: Sounds nice to me

**Comment by haikyuu** [comment]
@semoal Is it safe to use rematch v2 before this PR is merged? I'm not using the select plugin?

**Comment by semoal** [comment]
> @semoal Is it safe to use rematch v2 before this PR is merged? I'm not using the select plugin?

Yes absoutly, v2 won't change we're just improving/fixing select plugin development experience on typescript. But v2 is out of bugs, at least for now.

So feel free to use the next version. :)

**Comment by haikyuu** [comment]
Thanks @semoal

**Comment by semoal** [comment]
Probably should be a good idea to release this @tianzhich so select Users can start working and give US feedback of whats is pending?

**Comment by tianzhich** [comment]
> Probably should be a good idea to release this @tianzhich so select Users can start working and give US feedback of whats is pending?

sure, we need to release this. Now only a few typing problems in tests. We need more feedbacks.

**Pull request merged**